### PR TITLE
CartoDB D3 as leaflet tiled layer

### PR DESCRIPTION
**Pull request by fdansv**
- [x] Listen for tileAdded event.
- [x] SVG tile generation and layout in Leaflet
- [x] Relevant query generation
- [x] Geometries drawn in tiles
- [x] Correct tiles position on zoom/new tiles loaded (#6)
- [x] Make viz function work (#7)
- [x] Decouple renderer from Leaflet (#8)

@javisantana @rochoa

### CODE REVIEW

File: src/leaflet_d3.js [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+	initialize: function (options) {
+		var self = this;
+		options = options || {};
+		L.Util.setOptions(this, options);
+	},
+
+	onAdd: function (map) {
+		this._map = map;
+		this.options.map = map;
+		this.renderer = new Renderer(this.options);
```
Comment (rochoa) [inline src/leaflet_d3.js:32]: Shouldn't you require `Renderer` to use it here?

File: src/renderer.js [hunk 2]
```diff
@@ -457,31 +418,31 @@ D3CartoDBLayer = L.Class.extend({
...
-        /*
-        f.on('mouseover', function() {
-          f.style(styleForSymbolizer(sym, 'shader_hover'));
-        });
-        f.on('mouseout', function() {
-          f.style(styleForSymbolizer(sym, 'shader'));
-        });
-        */
-      }
+      // if (layer.hover) {
```
Comment (rochoa) [inline src/renderer.js]: I would remove this until we decide how this is gonna work.

File: src/leaflet_d3.js (hunk 1 above)
Comment (fdansv) [inline src/leaflet_d3.js:32]: Argh, yeah, that's using a global variable instead of require. Will change

File: src/renderer.js (hunk 2 above)
Comment (fdansv) [inline src/renderer.js]: :+1:

File: src/leaflet_d3.js (hunk 1 above)
Comment (fdansv) [inline src/leaflet_d3.js:32]: (changed :+1: )

**Comment by fdansv** [comment]
@rochoa @javisantana this is ready to merge on my side. I've opened #9 for tests. Should we do them/include them in this PR or merge and keep going?

File: src/leaflet_d3.js [hunk 3]
```diff
@@ -0,0 +1,217 @@
+var Renderer = require("./renderer");
+
+L.CartoDBd3Layer = L.Class.extend({
+
+	includes: [L.Mixin.Events, L.Mixin.TileLoader],
```
Comment (javisantana) [inline src/leaflet_d3.js:5]: I would actually use the tileloader mixing instead of copy the code here

File: src/renderer.js [hunk 4]
```diff
@@ -332,47 +203,56 @@ D3CartoDBLayer = L.Class.extend({
         .style(styleForSymbolizer(sym, 'shader'))
     }
   },
+  latLngToLayerPoint: function(lat, lng){
+    if(this._map._leaflet_id) {
```
Comment (javisantana) [inline src/renderer.js]: This should be a virtual function and should be implemented for each layer

File: src/renderer.js [hunk 5]
```diff
@@ -9,106 +9,20 @@ d3.selection.prototype.moveToFront = function() {
...
-*/
+var Renderer = function(options) {
+  this.options = options;
+  if (options.cartocss){
+    this.setCartoCSS(options.cartocss);
   }
+  this.collection = null;
+  this.globalVariables = {}
+  this.user = options.user;
+  this.sql_api_template = options.sql_api_template || 'http://{user}.cartodb.com';
```
Comment (javisantana) [inline src/renderer.js:19]: I'd start to think in data providers (fine for the moment)

File: src/leaflet_d3.js (hunk 3 above)
Comment (fdansv) [inline src/leaflet_d3.js:5]: Done :+1:

**Pull request merged**